src/crimson/osd/recovery_backend.h

```diff
@@ -150,7 +150,8 @@ class RecoveryBackend {
 	std::forward<F>(fut)
       ).finally([ref] {});
     }
-    seastar::future<> wait_for_recovered(BlockingEvent::TriggerI& trigger) {
+    template <typename T>
+    seastar::future<> wait_for_recovered(T &trigger) {
```

Comment:
OK, so when it's called from `start_replica_recovery_ops()` or `start_primary_recovery_ops()` the trigger's type becomes `AggregateBlockingEvent<WaitForObjectRecovery::BlockingEvent>::TriggerI`. The part triggers is created there:

```cpp
template <class T>
struct AggregateBlockingEvent {
  struct TriggerI {
    template <class FutureT>
    auto maybe_record_blocking(FutureT&& fut,
                               const typename T::Blocker& blocker) {
      // AggregateBlockingEvent is supposed to be used on relatively cold
      // paths (recovery), so we don't need to worry about the dynamic
      // polymothps / dynamic memory's overhead.
      return create_part_trigger()->maybe_record_blocking(
        std::move(fut), blocker);
    }
```